From 0b6e93398463bba2a6d3513e0161c399a0376ea7 Mon Sep 17 00:00:00 2001
From: Xaver Hugl <xaver.hugl@kde.org>
Date: Fri, 16 Oct 2026 16:06:17 +0200
Subject: [PATCH] opengl/gltexture: use swizzles to ignore non-existent alpha
 channels

QImage::convertTo doesn't set the "X" channel to all ones, so OpenGL can still
end up reading alpha values where there should be none. The swizzles make it
ignore the broken alpha channel.

Edited by fvogt@suse.de to also set swizzle in upload.
---
 src/opengl/gltexture.cpp | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/src/opengl/gltexture.cpp b/src/opengl/gltexture.cpp
index dbead12ec1..a650a024ac 100644
--- a/src/opengl/gltexture.cpp
+++ b/src/opengl/gltexture.cpp
@@ -218,6 +218,8 @@ void GLTexture::update(const QImage &image, const Region &region, const QPoint &
     glPixelStorei(GL_UNPACK_SKIP_PIXELS, 0);
     glPixelStorei(GL_UNPACK_SKIP_ROWS, 0);
 
+    setSwizzle(GL_RED, GL_GREEN, GL_BLUE, image.hasAlphaChannel() ? GL_ALPHA : GL_ONE);
+
     unbind();
 }
 
@@ -541,7 +543,14 @@ std::unique_ptr<GLTexture> GLTexture::upload(const QImage &image)
     glPixelStorei(GL_UNPACK_ROW_LENGTH, 0);
     glBindTexture(GL_TEXTURE_2D, 0);
 
-    return std::unique_ptr<GLTexture>(new GLTexture(GL_TEXTURE_2D, texture, internalFormat, image.size(), 1, true, OutputTransform::FlipY));
+    auto ret = std::unique_ptr<GLTexture>(new GLTexture(GL_TEXTURE_2D, texture, internalFormat, image.size(), 1, true, OutputTransform::FlipY));
+    if (!image.hasAlphaChannel()) {
+        ret->bind();
+        ret->setSwizzle(GL_RED, GL_GREEN, GL_BLUE, image.hasAlphaChannel() ? GL_ALPHA : GL_ONE);
+        ret->unbind();
+    }
+
+    return ret;
 }
 
 std::unique_ptr<GLTexture> GLTexture::upload(const QPixmap &pixmap)
-- 
2.56.0

